VirtoCommerce.CustomerModule.Data/Handlers/LogChangesMemberChangedEventHandler.cs

```diff
@@ -0,0 +1,204 @@
+    {
+        private readonly IChangeLogService _changeLogService;
+        private readonly string[] _observedProperties;
+
+        public static string[] ObservedProperties()
+        {
+            var memberPropsNames = ReflectionUtility.GetPropertyNames<Contact>(
+                x => x.Name, x => x.FirstName, x => x.LastName, x => x.MiddleName,
+                x => x.Salutation, x => x.FullName, x => x.BirthDate);
+            return memberPropsNames.ToArray();
+        }
+
+        public LogChangesMemberChangedEventHandler(IChangeLogService changeLogService)
+            : this(changeLogService, ObservedProperties())
+        {
+        }
+
+        protected LogChangesMemberChangedEventHandler(IChangeLogService changeLogService, string[] observedProperties)
+        {
+            _changeLogService = changeLogService;
+            _observedProperties = observedProperties;
+        }
+
+        public virtual Task Handle(MemberChangedEvent message)
+        {
+            var operationLogs =
+                message.ChangedEntries
+                    .Where(x => x.EntryState == EntryState.Modified && x.OldEntry is Contact)
+                    .SelectMany(GetChangedEntryOperationLogs)
+                    .ToArray();
+            _changeLogService.SaveChanges(operationLogs);
+            return Task.CompletedTask;
+        }
+
+        protected virtual IEnumerable<OperationLog> GetChangedEntryOperationLogs(
+            GenericChangedEntry<Member> changedEntry)
+        {
+            var result = new List<string>();
+
+            var diff = Comparer.Compare(changedEntry.OldEntry, changedEntry.NewEntry);
+
+            if (changedEntry.OldEntry is Contact contact)
+            {
+                result.AddRange(GetContactChanges(contact, changedEntry.NewEntry as Contact));
+                diff.AddRange(Comparer.Compare(contact, changedEntry.NewEntry as Contact));
+            }
+
+            var observedDifferences =
+                diff.Join(_observedProperties, x => x.Name.ToLowerInvariant(), x => x.ToLowerInvariant(), (x, y) => x)
+                    .ToArray();
+            foreach (var difference in observedDifferences.Distinct(new DifferenceComparer()))
+            {
+                AddMemberPropertyChangedOperation(result, changedEntry, difference);
+            }
+
+            return result.Select(x => GetLogRecord(changedEntry.NewEntry, x));
+        }
+
+        protected virtual IEnumerable<string> GetContactChanges(Contact originalContact, Contact modifiedContact)
+        {
+            return
+                GetAddressChanges(
+                    originalContact, originalContact.Addresses, modifiedContact.Addresses)
+                    .Union(GetEmailChanges(originalContact, originalContact.Emails, modifiedContact.Emails))
+                    .Union(GetPhoneChanges(originalContact, originalContact.Phones, modifiedContact.Phones));
+        }
+
+        protected virtual IEnumerable<string> GetPhoneChanges(
```

Comment:
I see code duplication in these methods: GetPhoneChanges, GetPhoneChanges, GetEmailChanges, GetAddressChanges can we do something with this? I think we can extend the generic GetListChanges method by new formal parameters and call it instead of these strongly typed methods. What do you think?

Reply:
We can inline them. It would look like this:
![image](https://user-images.githubusercontent.com/3872761/49639455-54421000-fa0b-11e8-8dc5-05f1c579f50f.png)
Not sure if it's better. `GetContactChanges` method is now very readable, with inlining it would be cumbersome